### PR TITLE
move composition velocity statistics from cookbook to main branch

### PR DESCRIPTION
**Pull request by elodie-kendall**
Pull Request Checklist. Please read and check each box with an X. Delete any part not applicable. Ask on the [forum](https://community.geodynamics.org/c/aspect) if you need help with any step.

After a discussion with Anne, I moved the postprocessor "composition velocity statistics" from the cookbook "kinematically_driven_subduction_2d" to the main branch as it's a useful feature for myself and some others. This way the user does not have to keep compiling the .so file etc.

### Before your first pull request:

* [ X] I have read the guidelines in our [CONTRIBUTING.md](../blob/main/CONTRIBUTING.md) document.

### For all pull requests:

* [X ] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [X ] I have tested my new feature locally to ensure it is correct.
* [ ] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [X ] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -443,7 +443,6 @@ LIST(APPEND CMAKE_REQUIRED_LIBRARIES ${CMAKE_DL_LIBS})
 CHECK_CXX_SOURCE_COMPILES("
 #include <cstddef>
 #include <dlfcn.h>
-
```
Comment (anne-glerum) [inline CMakeLists.txt:446]: Please undo changes to this file, I don't think they are related to this PR.

**Review by anne-glerum — COMMENTED** [review]
I haven't reviewed the changes, I just wanted to mention one thing.

**Comment by elodie-kendall** [comment]
I'm not sure why this is failing and the changes-test-results-master.diff is empty. Could someone advise me please? Thanks!

File: doc/modules/changes/20220729_ekendall [hunk 2]
```diff
@@ -0,0 +1,3 @@
+Changed: Moved composition velocity statistics script from cookbook to main code. I also changed the parameter terminology from "slab" to the more general name "selected_field".
```
Comment (anne-glerum) [inline doc/modules/changes/20220729_ekendall:1]:
```suggestion
Changed: Moved composition velocity statistics postprocessor from cookbook to main code. 
Compatibility of the input parameter is broken by renaming "slab" to the more general "selected_field". 
```

File: include/aspect/postprocess/composition_velocity_statistics.h [hunk 3]
```diff
@@ -66,7 +66,7 @@ namespace aspect
          */
 
       private:
-        std::vector<std::string> slab_compositions;
+        std::vector<std::string> selected_field_compositions;
```
Comment (anne-glerum) [inline include/aspect/postprocess/composition_velocity_statistics.h:69]: This name is a bit confusing. Several 'compositional fields' can make up an entity of which the rms velocity is to be computed. Maybe call it 'selected_fields' or 'selected_compositional_fields'?

File: source/postprocess/composition_velocity_statistics.cc [hunk 4]
```diff
@@ -81,23 +81,23 @@ namespace aspect
       std::vector<double> global_area_integral(local_area_integral.size());
       Utilities::MPI::sum(local_area_integral, this->get_mpi_communicator(), global_area_integral);
 
-      // compute the RMS velocity for each compositional field and for the slab
+      // compute the RMS velocity for each compositional field and for the selected compositonal field
```
Comment (anne-glerum) [inline source/postprocess/composition_velocity_statistics.cc:84]:
```suggestion
      // compute the RMS velocity for each compositional field and for the selected compositional fields combined
```

File: source/postprocess/composition_velocity_statistics.cc [hunk 5]
```diff
@@ -81,23 +81,23 @@ namespace aspect
...
       std::vector<double> vrms_per_composition(local_area_integral.size());
-      double velocity_square_integral_slab = 0., area_integral_slab = 0.;
+      double velocity_square_integral_selected_field = 0., area_integral_selected_field = 0.;
```
Comment (anne-glerum) [inline source/postprocess/composition_velocity_statistics.cc:86]:
```suggestion
      double velocity_square_integral_selected_fields = 0., area_integral_selected_fields = 0.;
```

Comment (anne-glerum) [reply source/postprocess/composition_velocity_statistics.cc:86]: Please update the names everywhere.

File: source/postprocess/composition_velocity_statistics.cc [hunk 6]
```diff
@@ -81,23 +81,23 @@ namespace aspect
...
             {
-              velocity_square_integral_slab += global_velocity_square_integral[c];
-              area_integral_slab += global_area_integral[c];
+              velocity_square_integral_selected_field += global_velocity_square_integral[c];
+              area_integral_selected_field += global_area_integral[c];
             }
         }
 
-      const double vrms_slab = std::sqrt(velocity_square_integral_slab) / std::sqrt(area_integral_slab);
+      const double vrms_selected_field = std::sqrt(velocity_square_integral_selected_field) / std::sqrt(area_integral_selected_field);
```
Comment (anne-glerum) [inline source/postprocess/composition_velocity_statistics.cc:100]:
```suggestion
      const double vrms_selected_fields = std::sqrt(velocity_square_integral_selected_fields) / std::sqrt(area_integral_selected_fields);
```

File: source/postprocess/composition_velocity_statistics.cc [hunk 7]
```diff
@@ -119,11 +119,11 @@ namespace aspect
             }
         }
 
-      // Also output the slab vrms
-      statistics.add_value("RMS velocity (" + unit + ") for slab ",
-                           time_scaling * vrms_slab);
+      // Also output the selected field vrms
```
Comment (anne-glerum) [inline source/postprocess/composition_velocity_statistics.cc:122]:
```suggestion
      // Also output the selected fields' vrms
```

Comment (anne-glerum) [reply source/postprocess/composition_velocity_statistics.cc:122]: Please update below accordingly

File: source/postprocess/composition_velocity_statistics.cc [hunk 8]
```diff
@@ -153,7 +153,7 @@ namespace aspect
       {
         prm.enter_subsection("Composition velocity statistics");
         {
-          prm.declare_entry("Names of slab compositional fields", "",
+          prm.declare_entry("Names of selected compositional fields", "",
```
Comment (anne-glerum) [inline source/postprocess/composition_velocity_statistics.cc:156]: The text of this parameter needs to be updated as well, i.e. 'the subducting plate' should be changed.

File: source/postprocess/composition_velocity_statistics.cc [hunk 9]
```diff
@@ -173,12 +173,12 @@ namespace aspect
...
-          slab_compositions = Utilities::split_string_list(prm.get("Names of slab compositional fields"));
+          selected_field_compositions = Utilities::split_string_list(prm.get("Names of selected compositional fields"));
 
-          AssertThrow((slab_compositions.size() > 0) &&
-                      (slab_compositions.size() <= this->n_compositional_fields()),
+          AssertThrow((selected_field_compositions.size() > 0) &&
+                      (selected_field_compositions.size() <= this->n_compositional_fields()),
                       ExcMessage("The length of the list of names for the compositional "
-                                 "fields that make up the slabs must be larger than zero "
+                                 "fields must be larger than zero "
```
Comment (anne-glerum) [inline source/postprocess/composition_velocity_statistics.cc:181]:
```suggestion
                                 "fields for which the RMS velocity is to be summed must be larger than zero "
```

**Review by anne-glerum — CHANGES_REQUESTED** [review]
Hi @elodie-kendall, I made some suggestions for the naming of the variables so that they better reflect what they are.

**Comment by elodie-kendall** [comment]
Thanks, I've addressed your comments

File: source/postprocess/composition_velocity_statistics.cc [hunk 10]
```diff
@@ -137,9 +137,9 @@ namespace aspect
                  << " " << unit;
           output << " // ";
         }
-      output << time_scaling *vrms_slab;
+      output << time_scaling *vrms_selected_fields;
 
-      return std::pair<std::string, std::string>("RMS velocity for compositions and slab:",
+      return std::pair<std::string, std::string>("RMS velocity for compositions and selected field:",
```
Comment (anne-glerum) [inline source/postprocess/composition_velocity_statistics.cc:142]:
```suggestion
      return std::pair<std::string, std::string>("RMS velocity for compositions and combined selected fields:",
```

File: source/postprocess/composition_velocity_statistics.cc [hunk 11]
```diff
@@ -153,10 +153,10 @@ namespace aspect
       {
         prm.enter_subsection("Composition velocity statistics");
         {
-          prm.declare_entry("Names of slab compositional fields", "",
+          prm.declare_entry("Names of selected compositional fields", "",
                             Patterns::List(Patterns::Anything()),
                             "A list of names for each of the compositional fields that "
-                            "makes up the subducting plate.");
+                            "you want to compute the RMS velocity for.");
```
Comment (anne-glerum) [inline source/postprocess/composition_velocity_statistics.cc:159]:
```suggestion
                            "you want to compute the combined RMS velocity for.");
```

**Review by anne-glerum — CHANGES_REQUESTED** [review]
Just 2 tiny comments. Now that this plugin is moved, we need a test that uses it as well. Can you provide one?

**Comment by anne-glerum** [comment]
Can someone else have a look at this as well?

**Comment by elodie-kendall** [comment]
Thanks for your last comments. I addressed them and added a test. I'll patch it after a final review.

**Review by tjhei — APPROVED** [review]

File: tests/kinematically_driven_subduction_2d_case1.prm [hunk 12]
```diff
@@ -0,0 +1,192 @@
+# This test is the same setup as the
+# cookbook kinematically_driven_subduction_2d
+# but with lower resolution. It models oceanic
```
Comment (tjhei) [inline tests/kinematically_driven_subduction_2d_case1.prm:3]:
```suggestion
# but with lower resolution. It models oceanic
```
Remove the space at the end of this line.

File: doc/modules/changes/20220729_ekendall [hunk 13]
```diff
@@ -0,0 +1,5 @@
+Changed: Moved composition velocity statistics postprocessor from cookbook to main code. 
+Compatibility of the postprocessor's input parameter "Names of slab compositional fields" is broken by renaming it to the more general "Names of selected compositional fields".
+Added: A test which is a replica of the cookbook "kinematically_driven_subduction_2d_case1" but with lower resolution and shorter end time.
```
Comment (anne-glerum) [inline doc/modules/changes/20220729_ekendall:3]: Please delete this line. Tests are added by default, so no need to report it.

**Review by anne-glerum — CHANGES_REQUESTED** [review]

**Comment by anne-glerum** [comment]
Hi Elodie, can you please delete the lines I mention above and patch the test results? Then this is ready to go!

**Review by anne-glerum — APPROVED** [review]

**Comment by elodie-kendall** [comment]
When I try to patch the test I get the error "Reversed (or previously applied) patch detected!  Assume -R? [n] n
" and if I press enter it says "Hunk #1 FAILED at 1.
Hunk #2 FAILED at 62.
2 out of 2 hunks FAILED"

**Comment by anne-glerum** [comment]
https://jenkins.tjhei.info/job/aspect/job/PR-4932/14/artifact/changes-test-results.diff

**Comment by anne-glerum** [comment]
Hi @elodie-kendall I'm sorry but this PR now includes changes to rheology and material model files. Can you remove those?

**Comment by anne-glerum** [comment]
Great, thanks for removing those files! There's one more empty line to remove with the diff:
https://jenkins.tjhei.info/job/aspect/job/PR-4932/16/artifact/changes-test-results.diff

**Comment by anne-glerum** [comment]
@tjhei the now failing test is unrelated (FPE), can this be ignored?

File: changes-test-results.diff [hunk 14]
```diff
@@ -0,0 +1,42 @@
+diff --git a/tests/kinematically_driven_subduction_2d_case1/screen-output b/tests/kinematically_driven_subduction_2d_case1/screen-output
```
Comment (gassmoeller) [inline changes-test-results.diff:1]: This file should not be added to the git history. Can you remove it by running `git rm ...` and then amend the commit (`git commit --amend`) and force push to this PR?

**Review by gassmoeller — COMMENTED** [review]
Just a small comment. Seems good otherwise.

**Review by gassmoeller — APPROVED** [review]
Thanks!

**Comment by elodie-kendall** [comment]
Thanks @gassmoeller and @anne-glerum

**Pull request merged**